Blaze: bail earlier when there are connection issues

jeherve opened the pull request:
Fixes #38068 

## Proposed changes:

When a site or user is not properly connected to WordPress.com, or when there are sync issues on the site, we can estimate that the site is not eligible to Blaze, since they won't be able to create Blaze campaigns when those things are not working properly.

Let's consequently bail earlier, consider Blaze as not ready to be initialized. That will save us a lot of failed remote requests later.


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

Related discussion: p1719416484308139-slack-CBG1CP4EN

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

This will be easier to test on a site with a broken Jetpack connection.

There, when the Blaze module is enabled in Jetpack > Settings > Traffic:

1. You should not see the navigation menu in Tools > Advertize.
2. You should not see any requests to WordPress.com being made.

jeherve committed:
Blaze: bail earlier when there are connection issues

When a site or user is not properly connected to WordPress.com, or when there are sync issues on the site, we can estimate that the site is not eligible to Blaze, since they won't be able to create Blaze campaigns when those things are not working properly.

Let's consequently bail earlier, consider Blaze as not ready to be initialized. That will save us a lot of failed remote requests later.

Related discussion: p1719416484308139-slack-CBG1CP4EN

jmdodd commented:
The broken site where I've encountered this condition passes both the connection and sync checks, but will still fail authentication. In this case, I suspect we should fall back to storing the failure in a transient.

jeherve committed:
Allow shortcircuiting as soon as the filter is set

If the filter is not set, we can go through the list of potential steps, one at a time, and shortcircuit whenever one condition is not met.

jeherve committed:
Store failed API requests for an hour

See https://github.com/Automattic/jetpack/pull/38066#issuecomment-2192196358

jeherve committed:
Add changelog

jeherve commented:
> In this case, I think we should fall back to storing the failure in a transient.

I implemented that in f418608ce2dba4a519761a4e9e01e2c2b2ebfc79

jeherve commented:
> The broken site where I encountered this condition passes both the connection and sync checks, but still fails authentication.

@Automattic/jetpack-vulcan Can you think of ways we could catch the errors @jmdodd described in p1719418092056169/1719416484.308139-slack-CBG1CP4EN , besides the existing `is_connected` and `is_user_connected` checks?

fgiannar commented:
> > The broken site where I encountered this condition passes both the connection and sync checks, but still fails authentication.
> 
> @Automattic/jetpack-vulcan Can you think of ways we could catch the errors @jmdodd described in p1719418092056169/1719416484.308139-slack-CBG1CP4EN , besides the existing `is_connected` and `is_user_connected` checks?

Those errors occur because we are attempting to access a private WoA site on an `wpcom/v2` endpoint on WPCOM using a blog token. 
We need to update the corresponding endpoint on WPCOM explicitly to allow requests signed with blog tokens.

Prior art: p1712319459536349/1706626329.415869-slack-C01U2KGS2PQ

jeherve committed:
Store transient as array to avoid confusion about status

See p1719485972517739/1719416484.308139-slack-CBG1CP4EN

fgiannar committed:
Handle both transient structures

This should avoid making new request right after plugin update.

fgiannar commented on projects/packages/blaze/src/class-blaze.php:
```diff
@@ -189,16 +199,27 @@ public static function site_supports_blaze( $blog_id ) {
 		}
 
+		// Allow short-circuiting the Blaze initialization via a filter.
+		if ( has_filter( 'jetpack_blaze_enabled' ) ) {
```
Is it intentional that we prioritized the filter over the Connection and Sync checks below? I would expect that even if the filter is used, we should still return `false` if there are connection issues or Sync is disabled.

jeherve replied:
In practice, the filter is really used in 2 scenarios:

- In tests
- In third-party plugins / themes' `functions.php` files as a way to force Blaze to false.

With that in mind, I think it's okay to prioritize the filter, since in real life I don't see folks use it to force Blaze to true?

jeherve commented:
> Those errors occur because we are attempting to access a private WoA site on an `wpcom/v2` endpoint on WPCOM using a blog token.
> We need to update the corresponding endpoint on WPCOM explicitly to allow requests signed with blog tokens.

Thank you! I just opened D153591-code to try to tackle this.

fgiannar reviewed (changes requested):
The `if ( ! is_array( $result ) || empty( $result['approved'] ) ) {` check in `class-blaze.php` line 175 will result in returning early and not caching the result if `$result['approved']` is false.
Maybe consider using `isset`?

jeherve committed:
Try adding a test for site_supports_blaze

It fails for me locally, but by pushing it I can share it and get feedback from others.

fgiannar committed:
Cache results even when the site is not eligible

See https://github.com/Automattic/jetpack/pull/38066#pullrequestreview-2145377103

jeherve commented:
> The `if ( ! is_array( $result ) || empty( $result['approved'] ) ) {` check in `class-blaze.php` line 175 will result in returning early and not caching the result if `$result['approved']` is false.
> Maybe consider using `isset`?

👍 Done in 0ff1eaa7f392d4d1982c5f652f06370751955a68

jeherve commented:
> Btw are there any existing unit tests for Blaze::site_supports_blaze ? Maybe we could add the transient logic there too

I worked on a test in 5045a01c710c93bdead36ed702e0f2fb86f68b96, but cannot get it to work. At this point I've spent too much time on it already, so I'm pushing my work in progress, and hoping someone can tell me what I did wrong :)

Thank you!

fgiannar committed:
Fix unit tests

fgiannar commented:
> At this point I've spent too much time on it already, so I'm pushing my work in progress, and hoping someone can tell me what I did wrong :)

The request using `Client::wpcom_json_api_request_as_blog` was failing early because 1) the [blog token was missing](https://github.com/Automattic/jetpack/blob/a954b88239c722671bc27fc02c3d9265af5a7923/projects/packages/connection/src/class-client.php#L114) and 2) the `host` part for the remote request [was missing](https://github.com/Automattic/jetpack/blob/a954b88239c722671bc27fc02c3d9265af5a7923/projects/packages/connection/legacy/class-jetpack-signature.php#L209)

Fixed via https://github.com/Automattic/jetpack/pull/38066/commits/54c72b86f3b8ec6619bc5a6607471e5493c71391

fgiannar reviewed (approved):
Great work here, Jer! Many thanks for these performance improvements for Blaze and adding a relevant unit test too 👍
